### DISCUSSION TITLE
Migrate from lb2 to lv4, convert timestamp string to date error:  timestamp string is not a valid date, it also appears in lb3

### DISCUSSION THREAD

**Discussion by Jackclarify** [loopbackio/loopback-next#7906] [LB3 Migration]
Env:
```
node:12.16.0
npm:6.13.7
postgresql:12.6.0
os:  OS Name	Microsoft Windows 10 Pro  Version 10.0.19043 Build 19043
"loopback": "3.28.0"
```

Database model:
```
    {
        "name": "SavedAnalysis",
        "replaceOnPUT": false,
        "base": "PersistedModel",
        "idInjection": false,
        "options": {
            "postgresql": {
                "table": "saved_analysis"
            },
            "authorizeForAll": [
                "read",
                "findById",
                "create"
            ]
        },
        "properties": {
            "id": {
                "type": "string",
                "id": true,
                "defaultFn": "uuid"
            },
            "title": {
                "type": "string"
            },
            "createdDate": {
                "type": "date",
                "postgresql": {
                    "columnName": "created_date"
                }
            },
            "seeded": {
                "type": "boolean"
            }
        },
        "validations": [],
        "relations": {},
        "acls": [],
        "methods": {}
    }
```

Request body: put
`{"model":"SavedAnalysis","data":{"title":"RF-A-13-SavedAnalysesSpotlights-1","createdDate":"1631951099611","authorEmail":"xxx@gmail.com"}}`

Method 
app = require('../../server/server'); <== lb3app server
app.models['SavedAnalysis'].upsert(data)

dependencies:
```
    "loopback": "^3.0.0",
    "loopback-boot": "^3.3.0",
    "loopback-component-explorer": "^6.5.0",
    "loopback-component-push": "^3.4.0",
    "loopback-connector": "https://github.com/clarifyhealth/clarify-loopback-connector.git#TVF", <== depend on version 2.3.0
    "loopback-connector-postgresql": "=2.8.0",
    "loopback-connector-remote": "^1.0.3",
    "loopback-connector-rest": "^2.0.0",
    "loopback-connector-twilio": "^0.5.1",
    "loopback-filters": "^0.1.3",
```

Error:
```
    {
        "error": {
            "statusCode": 422,
            "name": "ValidationError",
            "message": "The `SavedAnalysis` instance is not valid. Details3333333333333333: `createdDate` is not a valid date (value: Invalid Date).",
            "details": {
                "context": "SavedAnalysis",
                "codes": {
                    "createdDate": [
                        "date"
                    ]
                },
                "messages": {
                    "createdDate": [
                        "is not a valid date"
                    ]
                }
            }
        }
    }
```

I found it might relate to [loopback-datasource-juggler model-builder lib](https://github.com/loopbackio/loopback-datasource-juggler/blob/v3.36.1/lib/model-builder.js#L697-L701)
```
function DateType(arg) {
  if (arg === null) return null;
  console.log(arg);  // <== 1631951099611 
  const d = new Date(arg);
  console.log(d); // <== Invalid Date, expect: 2021-09-18T07:44:59.611Z
  return d;
}
```
Do we rewrite the Date method? It works well when I remove new keyword.

Can anyone give me some hints?
Thanks!

@bajtos @rmg @achrinza @emazzu @raymondfeng

**Comment by collaorodrigo7** [comment]
I am not 100% sure if it's the right option, but I think an interceptor can help you in this case.

[Here](https://loopback.io/doc/en/lb4/Interceptor.html#example-interceptors) in number 3 there is an example of how to use an interceptor to change a string to upper case. 

Also, I created a repo with a sample of how you could use the interceptor to fix your date issue:
https://github.com/collaorodrigo7/loopback4-validator-sample

However, I was not able to keep the CRUD validations, as those will execute before the interceptor, causing it to return an error. [See here](https://github.com/collaorodrigo7/loopback4-validator-sample/blob/main/src/controllers/book.controller.ts#L46)

@dhmlau what do you think? Is this a good use case of an interceptor? or what would you recommend? Also, Is there any way we could make the interceptor execute before the CRUD validation (from the requestBody decorator)?

**Comment by Xiaohang0316** [comment]
It seems more convenient to create a dateFormat method for processing

>

```
dateFormat(date) {
        if ( date &&  !_.toString(date).includes('-')) {
            date = new Date(_.toNumber(date));
        }
       return date
}
```
